### PR TITLE
Convert variable types before applying defaults

### LINKED ISSUE #31978: "inconsistent set element types" crash
### Terraform Version

```shell
Terraform v1.3.2
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v4.34.0
```

### Terraform Configuration Files

```terraform

variable "backup_plans" {
  type = set(object({
    name = string
    schedules = set(object({
      name                     = string
      cron_expression          = string
      retention_days           = number
      cold_storage_after       = optional(number, 0)
      enable_continuous_backup = bool
    }))
    conditions          = set(map(string))
    recovery_point_tags = map(string)
    resources           = list(string)
    not_resources       = list(string)
  }))
default = [
    {
      name = "test"
      schedules = [
        {
          name                     = "daily"
          cron_expression          = "cron(0 1 * * ? *)"
          retention_days           = 7
          enable_continuous_backup = true
        },
        {
          name                     = "weekly"
          cron_expression          = "cron(0 3 ? * MON *)"
          retention_days           = 13 * 7
          cold_storage_after       = 1
          enable_continuous_backup = false
        }
      ]
      conditions = [
        {
          key   = "datatype"
          value = "business-critical"
        }
      ]
      not_resources = []
      resources = [
        "arn:aws:dynamodb:*:*:table/*"
      ]

      recovery_point_tags = {
        "backuptype" = "fallback"
      }
    },

  ]
}

resource "aws_backup_plan" "this" {
  for_each = { for plan in var.backup_plans : plan.name => plan }
  name     = each.key

  dynamic "rule" {
    for_each = each.value.schedules
    content {
      rule_name                = rule.value.name
      target_vault_name        = aws_backup_vault.this.name
      schedule                 = rule.value.cron_expression
      enable_continuous_backup = rule.value.enable_continuous_backup
      lifecycle {
        delete_after       = rule.value.retention_days
        cold_storage_after = rule.value.cold_storage_after
      }
      recovery_point_tags = each.value.recovery_point_tags
    }
  }
}


resource "aws_backup_selection" "this" {
  for_each     = { for plan in var.backup_plans : plan.name => plan }
  iam_role_arn = aws_iam_role.this.arn
  name         = "default"
  plan_id      = aws_backup_plan.this[each.key].id
  condition {
    dynamic "string_equals" {
      for_each = each.value.conditions
      content {
        key   = "aws:ResourceTag/${string_equals.value.key}"
        value = string_equals.value.value
      }
    }
  }
  not_resources = each.value.not_resources
  resources     = each.value.resources
}

```

### Debug Output

.

### Expected Behavior

no crash

### Actual Behavior

inconsistent set element types (cty.Object(map[string]cty.Type{"conditions":cty.Tuple([]cty.Type{cty.Object(map[string]cty.Type{"key":cty.String, "value":cty.String})}), "name":cty.String, "not_resources":cty.List(cty.String), "recovery_point_tags":cty.Object(map[string]cty.Type{"ts-backuptype":cty.String}), "resources":cty.Tuple([]cty.Type{cty.String}), "schedules":cty.List(cty.Object(map[string]cty.Type{"cold_storage_after":cty.Number, "cron_expression":cty.String, "enable_continuous_backup":cty.String, "name":cty.String, "retention_days":cty.String}))}) then cty.Object(map[string]cty.Type{"conditions":cty.Tuple([]cty.Type{cty.Object(map[string]cty.Type{"key":cty.String, "value":cty.String})}), "name":cty.String, "not_resources":cty.List(cty.String), "recovery_point_tags":cty.Object(map[string]cty.Type{"ts-backuptype":cty.String}), "resources":cty.Tuple([]cty.Type{cty.String}), "schedules":cty.List(cty.Object(map[string]cty.Type{"cold_storage_after":cty.String, "cron_expression":cty.String, "enable_continuous_backup":cty.String, "name":cty.String, "retention_days":cty.String}))}))
goroutine 31 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x1c
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/Users/distiller/project/project/internal/logging/panic.go:55 +0x170
panic({0x104c352e0, 0x1400149ad20})
	/usr/local/go/src/runtime/panic.go:884 +0x204
github.com/zclconf/go-cty/cty.SetVal({0x140031d2280, 0x4, 0x1?})
	/Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.11.0/cty/value_init.go:281 +0x608
github.com/zclconf/go-cty/cty.transform({0x0?, 0x0, 0x0}, {{{0x105039f78?, 0x14001479280?}}, {0x104f49d60?, 0x14003856a50?}}, {0x105021360, 0x14000634390})
	/Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.11.0/cty/walk.go:177 +0xbe0
github.com/zclconf/go-cty/cty.TransformWithTransformer(...)
	/Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.11.0/cty/walk.go:130
github.com/hashicorp/hcl/v2/ext/typeexpr.(*Defaults).Apply(0x140008297e0?, {{{0x105039f78?, 0x14001479280?}}, {0x104f49d60?, 0x14003856a50?}})
	/Users/distiller/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.14.1/ext/typeexpr/defaults.go:38 +0x9c
github.com/hashicorp/terraform/internal/terraform.prepareFinalInputVariableValue({{0x14000ab2a60, 0x1, 0x1}, {{}, {0x1400039eaf0, 0xc}}}, 0x14003f15a30, 0x1400077a750)
	/Users/distiller/project/project/internal/terraform/eval_variable.go:98 +0x88c
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).evalModuleVariable(0x140036324e0?, {0x10504e2f8?, 0x140023ca1c0?}, 0xc?)
	/Users/distiller/project/project/internal/terraform/node_module_variable.go:239 +0x3e4
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).Execute(0x140036324e0, {0x10504e2f8, 0x140023ca1c0}, 0x2)
	/Users/distiller/project/project/internal/terraform/node_module_variable.go:155 +0x100
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0x1400465e870, {0x10504e2f8, 0x140023ca1c0}, {0x1304d1a38, 0x140036324e0})
	/Users/distiller/project/project/internal/terraform/graph_walk_context.go:136 +0xa8
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x104e04d40, 0x140036324e0})
	/Users/distiller/project/project/internal/terraform/graph.go:74 +0x208
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0x14003632540, {0x104e04d40, 0x140036324e0}, 0x140010e6a80)
	/Users/distiller/project/project/internal/dag/walk.go:381 +0x2e0
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update
	/Users/distiller/project/project/internal/dag/walk.go:304 +0xbf0

### Steps to Reproduce

terraform plan

### Additional Context

The issue only happens if cold_storage_after is set in a schedule. 

### References

_No response_

### PR DESCRIPTION
**Pull request by liamcervante**
Fixes #31978

Currently Terraform attempts to apply defaults in variables before calling go-cty to convert the variables into their actual concrete types. This can lead to a situation where HCL cannot apply the defaults because of a type incompatibility and then panics and crashes Terraform (#31978). It's kind of impossible to recover from or predict this situation without modifying the HCL public functions (eg. make the panic return an error, or add a new function like `CheckCanApplyDefaults`). We can't do this check with go-cty because as far as go-cty is concerned these are perfectly valid conversions. Note, that HCL and go-cty disagree here about what is or what isn't a valid type conversion.

This PR switches the order of these operations. We now go through go-cty to convert to the concrete type first, then when we call the apply defaults function the types are guaranteed to match and HCL doesn't need to worry about any kind of conversions.

This change requires an upstream fix in go-cty, so go-cty is updated as part of this change.

## Target Release

1.3.4

### BUG FIXES

-  Input and Module Variables: Convert variable types before attempting to apply default values

### CODE REVIEW

**Comment by alisdair** [comment]
I'm not sure this change is valid, but I'm also not sure it's invalid.

We originally applied defaults before converting type in part to allow us to distinguish between missing object attributes and attributes with `null` value. This distinction was later removed in order to harmonize the treatment of `null` with other contexts, so it may no longer be necessary.

However, you should be able to assign a `map` value to an `object` variable, so the existing test that you changed surprises me. From glancing at the code I'm not immediately sure why that would have broken, but I'll take another look.

**Comment by liamcervante** [comment]
> why that would have broken

It breaks because when it arrives in the `convert` package it's a map of strings. The convert package tries to turn it into the object but because one of the attributes of the object is a complex object that is not convertable from a string the whole convert attempt fails. This is different to the behaviour of inserting defaults which is quite happy to convert a map of strings into the complex object type as it is going through putting the correct defaults in.

I did have a look at going into go-cty and fixing the convert function to allow this. Here's a test case I wrote in go-cty that matches what the test case here is essentially trying to do: https://github.com/zclconf/go-cty/pull/139.

**Comment by liamcervante** [comment]
I've reverted the change I had to make to the other test, which means the tests are now going to report as failing. But once the following PRs are merged into go-cty and released this test will start passing again:

- https://github.com/zclconf/go-cty/pull/140
- https://github.com/zclconf/go-cty/pull/139
- https://github.com/zclconf/go-cty/pull/141

**Comment by liamcervante** [comment]
This is ready for review!

File: internal/terraform/eval_variable.go [hunk 1]
```diff
@@ -140,6 +132,14 @@ func prepareFinalInputVariableValue(addr addrs.AbsInputVariableInstance, raw *In
 		return cty.UnknownVal(cfg.Type), diags
 	}
 
+	// Apply defaults from the variable's type constraint to the given value,
+	// unless the given value is null. We do not apply defaults to top-level
```
Comment (alisdair) [inline internal/terraform/eval_variable.go:136]:
```suggestion
	// Apply defaults from the variable's type constraint to the converted value,
	// unless the converted value is null. We do not apply defaults to top-level
```

The use of the word "given" here was intentionally referring to the `given` variable, which is unconverted. The suggestion in this comment is one way to address the issue I have with this comment, which is that I think applying defaults after conversion is a bit subtle and we should try to leave a note about it.

**Review by alisdair — APPROVED** [review]

**Pull request merged**